sync/src/main/java/tech/pegasys/artemis/sync/SyncManager.java

```diff
@@ -0,0 +1,94 @@
+import com.google.common.primitives.UnsignedLong;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.networking.eth2.Eth2Network;
+import tech.pegasys.artemis.networking.eth2.peers.Eth2Peer;
+import tech.pegasys.artemis.networking.eth2.rpc.core.ResponseStream;
+import tech.pegasys.artemis.statetransition.StateTransition;
+import tech.pegasys.artemis.statetransition.StateTransitionException;
+import tech.pegasys.artemis.storage.ChainStorageClient;
+import tech.pegasys.artemis.storage.Store;
+import tech.pegasys.artemis.storage.events.StoreDiskUpdateEvent;
+
+public class SyncManager {
+
+  private final Eth2Network network;
+  private final ChainStorageClient storageClient;
+  private final StateTransition stateTransition;
+  private final EventBus eventBus;
+
+  private Set<Eth2Peer> peerBlacklist;
+
+  public SyncManager(
+      final Eth2Network network, final ChainStorageClient storageClient, final EventBus eventBus) {
+    this.network = network;
+    this.storageClient = storageClient;
+    this.eventBus = eventBus;
+    this.stateTransition = new StateTransition(false);
+    this.peerBlacklist = new HashSet<>();
+  }
+
+  private void sync() {
+    Eth2Peer syncPeer = findSyncPeer().orElseThrow();
+    requestSyncBlocks(
+        syncPeer,
+        ((block) -> {
+          try {
+            Store.Transaction transaction = storageClient.getStore().startTransaction();
+            on_block(transaction, block, stateTransition);
+            transaction.commit();
+            eventBus.post(new StoreDiskUpdateEvent(transaction));
+          } catch (StateTransitionException e) {
+            peerBlacklist.add(syncPeer);
+            sync();
+          }
+        }));
+  }
+
+  private Optional<Eth2Peer> findSyncPeer() {
+    UnsignedLong ourFinalizedEpoch = storageClient.getStore().getFinalizedCheckpoint().getEpoch();
+    return network
+        .streamPeers()
+        .filter(peer -> !peerBlacklist.contains(peer))
+        .filter(
+            peer -> {
+              Eth2Peer.StatusData statusData = peer.getStatus();
+              return statusData.getFinalizedEpoch().compareTo(ourFinalizedEpoch) > 0;
+            })
+        .findAny();
+  }
+
+  private void requestSyncBlocks(
+      Eth2Peer peer, ResponseStream.ResponseListener<BeaconBlock> blockResponseListener) {
+    Eth2Peer.StatusData peerStatusData = peer.getStatus();
+    Bytes32 headBlockRoot = peerStatusData.getHeadRoot();
+    UnsignedLong headBlockSlot = peerStatusData.getHeadSlot();
+    UnsignedLong startSlot =
+        compute_start_slot_at_epoch(storageClient.getStore().getFinalizedCheckpoint().getEpoch());
+    UnsignedLong step = UnsignedLong.ONE;
+    UnsignedLong count = headBlockSlot.minus(startSlot);
+    peer.requestBlocksByRange(headBlockRoot, startSlot, count, step, blockResponseListener);
```

Comment:
I agree. Syncing until the finalized epoch of the peer would be a more robust target. It would also allow us to cleanly separate syncing the historical state and the recent state as @mbaxter previously suggested.

Reply:
Actually, I'm taking this back. According to the networking specification, even if there is a peer re-orgs, the node has to still complete the request. 

> Clients MUST support head_block_root values since the latest finalized epoch.

So in that scenario, even if the headBlockRoot we initially received is not in their new re-orged chain, the chain defined by the initial headBlockRoot should be finalizing blocks up to their higher finalized epoch that they initially advertised, and we should still sync to them, and they should be providing us blocks regardless.

Reply:
I wouldn't trust anything the spec says to be honest. There are a few places it requires sending impossible responses.

Certainly in the discussions at interop it was thought that this was a check to ensure we were on the right chain so the request would be rejected otherwise.  We should confirm what other clients actually do before making any strong assumptions here - although this probably isn't a detail we need to sort out for a very minimal sync implementation.